### PR TITLE
[Feature](bangc-ops): add strided_slice binary operator.

### PR DESCRIPTION
**Pull request by zhengleiZL**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

将strided_slice算子以二进制形式合入mluops

## 2. Modification

增加算子strided_slice

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff3
    - [x] mlu diff3 == 0
    - [x] mlu diff3_1 == 0
    - [x] mlu diff3_2 == 0

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [x] MLU370
  - [x] MLU590
- Job types
  - [ ] BLOCK
  - [ ] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test (e.g. float32/int8)
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

#### 3.2.2 Parameter Check

Test Point-1: `When a new operator is submitted, the test points are given and the test results are stated`. Acceptance Standard: `Normal error`.
```bash
Please fill your test results(Error Message) in here, ...
```

Test Point-2: `Whether illegal parameters are passed`. Acceptance Standard: `Normal error`.
```bash
Test results...
```

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370
```bash
[----------] Global test environment tear-down
[ SUMMARY  ] Total 1 cases of 1 op(s).
ALL PASSED.
[==========] 1 test case from 1 test suite ran. (670 ms total)
[  PASSED  ] 1 test case.
```

Platform：MLU590
```bash
[----------] Global test environment tear-down
[ SUMMARY  ] Total 1 cases of 1 op(s).
ALL PASSED.
[==========] 1 test case from 1 test suite ran. (1107 ms total)
[  PASSED  ] 1 test case.
```

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -15093,6 +15093,95 @@ mluOpSyncBatchNormBackwardElemtV2(mluOpHandle_t handle,
...
+ *   Input. The descriptor of the output tensor. For detailed information,
+ *   see ::mluOpTensorDescriptor_t.
+ * @param[out]  output
+ *   Output. Pointer to the MLU memory that stores the output tensor.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM, ::MLUOP_STATUS_NOT_SUPPORTED,
+ *   ::MLUOP_STATUS_INTERNAL_ERROR
+ *
+ * @par Formula
+ * - See "StridedSlice Operation" section in "Cambricon MLUOP User Guide" for details.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15127]: 再确认一下，用户手册里这个章节没写公式呀

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -15093,6 +15093,95 @@ mluOpSyncBatchNormBackwardElemtV2(mluOpHandle_t handle,
...
+ *
+ * @par Data Type
+ * - The (I/O)function supports the following byte-width data types for \p input and \p output tensors.
+ *   The byte width of a data type can be got with the ::mluOpGetSizeOfDataType function.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15131]:
```suggestion
 *   The byte width of a data type can be obtained by ::mluOpGetSizeOfDataType function.
```

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -15093,6 +15093,95 @@ mluOpSyncBatchNormBackwardElemtV2(mluOpHandle_t handle,
...
+ *   Note that the data type of input tensor \p input and output tensor \p output must be the same.
+ *   The supported data types are as follows:
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15133]: 这里再确认一下哦，下文是data type么？字符宽度吧

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -15093,6 +15093,95 @@ mluOpSyncBatchNormBackwardElemtV2(mluOpHandle_t handle,
...
+ *   - input tensor: 1-byte, 2-byte, 4-byte, 8-byte.
+ *   - output tensor: 1-byte, 2-byte, 4-byte, 8-byte.
+ *
+ * @par Scale Limitation
+ * - The input parameters should meet
+ *   the following requirements (dim0 means the first dimension of input, dim0_stride
+ *   means stride of dim0, dim0_input means the value of dim0 dimension of input,
+ *   dim0_begin is means beginning index of dim0, dim0_end means ending index of dim0):
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15141]: 不建议用括号的形式，括号中内容提取出来放到最前面or 最后面做注释吧

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -15093,6 +15093,95 @@ mluOpSyncBatchNormBackwardElemtV2(mluOpHandle_t handle,
...
+ *   - dim0_stride * dim1_stride * dim2_stride * dim3_stride != 0.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15142]:
```suggestion
 *   - ``dim0_stride * dim1_stride * dim2_stride * dim3_stride != 0``.
```

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -15093,6 +15093,95 @@ mluOpSyncBatchNormBackwardElemtV2(mluOpHandle_t handle,
...
+ *   - When dim0_stride > 0, (dim0_begin >= 0 && dim0_end > dim0_begin
+ *   && dim0_input >= dim0_end) should be true.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15144]: should改成must吧

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/strided_slice/strided_slice.cpp [hunk 7]
```diff
@@ -0,0 +1,194 @@
+/*************************************************************************
+ * Copyright (C) [2019-2022] by Cambricon, Inc.
```
Comment (AquamarineSu) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/strided_slice/strided_slice.cpp:2]: 时间是不是改成2023

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -15093,6 +15093,95 @@ mluOpSyncBatchNormBackwardElemtV2(mluOpHandle_t handle,
                                   const void *count,
                                   const mluOpTensorDescriptor_t diff_x_desc,
                                   void *diff_x);
+
+// Group:StridedSlice
+/*!
+ * @brief Slices on input tensor \p input with the begin \p begin, end \p end,
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15099]: 这句描述是不是有点儿简单，很难看出具体操作，等我找几个其他模块的示例有度发你~

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -15093,6 +15093,95 @@ mluOpSyncBatchNormBackwardElemtV2(mluOpHandle_t handle,
...
+ * stride \p stride, and returns the results in the output tensor \p output.
+ *
+ * @param[in]  handle
+ *   Input. Handle to a Cambricon MLUOP context that is used to manage MLU devices and
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15103]:
```suggestion
 *   Handle to a Cambricon MLUOP context that is used to manage MLU devices and
```

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -15093,6 +15093,95 @@ mluOpSyncBatchNormBackwardElemtV2(mluOpHandle_t handle,
...
+ *   queues in the strided_slice operation. For detailed information,
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15104]:
```suggestion
 *   queues. For detailed information,
```

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -15093,6 +15093,95 @@ mluOpSyncBatchNormBackwardElemtV2(mluOpHandle_t handle,
...
+ *   Input. An array that stores the starting position of each dimension of input to be sliced.
+ * @param[in]  end
+ *   Input. An array that stores the ending position of each dimension of input to be sliced.
+ * @param[in]  stride
+ *   Input. An array that stores the stride of each dimension of input to be sliced.
+ * @param[in]  output_desc
+ *   Input. The descriptor of the output tensor. For detailed information,
+ *   see ::mluOpTensorDescriptor_t.
+ * @param[out]  output
+ *   Output. Pointer to the MLU memory that stores the output tensor.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15121]: 同上， 出参和入参已经使用中括号标注了，参数解释无需重复

File: bangc-ops/mlu_op.h (hunk 1 above)
Comment (zhengleiZL) [reply bangc-ops/mlu_op.h:15128]: 已增加

File: bangc-ops/mlu_op.h (hunk 2 above)
Comment (zhengleiZL) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 3 above)
Comment (zhengleiZL) [reply bangc-ops/mlu_op.h]: 是字符宽度，重新修改上述描述

File: bangc-ops/mlu_op.h (hunk 6 above)
Comment (zhengleiZL) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/strided_slice/strided_slice.cpp (hunk 7 above)
Comment (zhengleiZL) [reply bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/strided_slice/strided_slice.cpp]: 修改为2023

File: bangc-ops/mlu_op.h (hunk 11 above)
Comment (zhengleiZL) [reply bangc-ops/mlu_op.h]: 已删除

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 12]
```diff
@@ -943,10 +943,29 @@ mluOpSyncBatchNormElemt
...
+
+mluOpStridedSlice
+-----------------------
+该算子从给定input张量中提取一个 ``(end - begin) / stride`` 的片段。从指定的位置开始 ``begin`` ，每隔 ``stride`` 个索引，直到所有维度都不小于 ``end`` 结束。当 ``stride`` 为负数时，会导致反向切片。公式如下：
+
+.. math::
+
+   output = input[begin[0]:end[0]:stride[0],
+                 ...,
+                 begin[N]:end[N]:stride[N]]
```
Comment (YunyiGUANG) [inline docs/bangc-docs/user_guide/9_operators/index.rst:964]: 这里是不是应该是N-1

**Review by YunyiGUANG — APPROVED** [review]

File: bangc-ops/mlu_op.h [hunk 13]
```diff
@@ -15093,6 +15093,99 @@ mluOpSyncBatchNormBackwardElemtV2(mluOpHandle_t handle,
...
+ *   ::MLUOP_STATUS_INTERNAL_ERROR
+ *
+ * @par Formula
+ * - See "StridedSlice Operation" section in "Cambricon MLUOP User Guide" for details.
+ *
+ * @par Data Type
+ * - The (I/O)function supports the following byte-width data types for \p input and \p output tensors.
+ *   The byte width of a data type can be obtained by ::mluOpGetSizeOfDataType function.
+ *   Note that the data type of input tensor \p input and output tensor \p output must be the same.
+ *   The supported byte-width data types are as follows:
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15134]: 列表的级别是不是调整一下

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 14]
```diff
@@ -943,10 +943,29 @@ mluOpSyncBatchNormElemt
...
+
+其中：
+
+- input为 ``N`` 个维度的张量
+- output为 ``N`` 个维度的张量
+- ``begin`` ， ``end`` 和 ``stride`` 均为长度为 ``N`` 的数组，分别表示每个维度切片的起始索引，结束索引和步幅。
```
Comment (AquamarineSu) [inline docs/bangc-docs/user_guide/9_operators/index.rst:970]: 句末加标点哦~

File: bangc-ops/mlu_op.h [hunk 15]
```diff
@@ -15093,6 +15093,99 @@ mluOpSyncBatchNormBackwardElemtV2(mluOpHandle_t handle,
                                   const void *count,
                                   const mluOpTensorDescriptor_t diff_x_desc,
                                   void *diff_x);
+
+// Group:StridedSlice
+/*!
+ * @brief Extracts a slice of size ``(end - begin) / stride`` from the given \p input tensor.
+ * Strating at the location specified by \p begin the slice continues by adding \p stride to
+ * the index until all dimensions are not less than \p end. Note that a \p stride can be
```
[the remark this answers is not in the document]
Comment (zhengleiZL) [reply bangc-ops/mlu_op.h:15101]: 已修改

File: bangc-ops/mlu_op.h (hunk 13 above)
Comment (zhengleiZL) [reply bangc-ops/mlu_op.h:15134]: 已调整

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/strided_slice/strided_slice.cpp [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#include <complex>
+#include "strided_slice.h"
```
Comment (PetrelYy) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/strided_slice/strided_slice.cpp:13]:
```suggestion
#include "strided_slice.h"

#include <complex>
```

File: bangc-ops/mlu_op.h [hunk 17]
```diff
@@ -15093,6 +15093,99 @@ mluOpSyncBatchNormBackwardElemtV2(mluOpHandle_t handle,
...
+ *   For detailed information, see ::mluOpHandle_t.
+ * @param[in]  input_desc
+ *   The descriptor of the input tensor. For detailed information,
+ *   see ::mluOpTensorDescriptor_t.
+ * @param[in]  input
+ *   Pointer to the MLU memory that stores the input tensor.
+ * @param[in]  begin
+ *   An array that stores the starting position of each dimension of input to be sliced.
+ * @param[in]  end
+ *   An array that stores the ending position of each dimension of input to be sliced.
```
Comment (PetrelYy) [inline bangc-ops/mlu_op.h:15115]: 缩进，格式参考 https://wiki.cambricon.com/pages/viewpage.action?pageId=255658083

File: bangc-ops/mlu_op.h [hunk 18]
```diff
@@ -15093,6 +15093,99 @@ mluOpSyncBatchNormBackwardElemtV2(mluOpHandle_t handle,
                                   const void *count,
                                   const mluOpTensorDescriptor_t diff_x_desc,
                                   void *diff_x);
+
+// Group:StridedSlice
+/*!
+ * @brief Extracts a slice of size ``(end - begin) / stride`` from the given \p input tensor.
+ * Strating at the location specified by \p begin the slice continues by adding \p stride to
+ * the index until all dimensions are greater than or equal to \p end. Note that a \p stride
```
Comment (Unireverse) [inline bangc-ops/mlu_op.h:15101]:
```suggestion
 * Strating at the location specified by \p begin, the slice continuously adds \p stride to
 * the index until all dimensions are greater than or equal to \p end. Note that a \p stride
```

Comment (Unireverse) [reply bangc-ops/mlu_op.h:15101]: 感觉这么改合理一些，问问文档组

Comment (zhengleiZL) [reply bangc-ops/mlu_op.h:15101]: 已修改

**Review by PetrelYy — APPROVED** [review]

**Review by AquamarineSu — APPROVED** [review]

**Review by AquamarineSu — APPROVED** [review]

**Pull request merged**